### PR TITLE
Compare the DC of the underlying type of a FriendDecl

### PR DESCRIPTION
**Pull request by martong**

### CODE REVIEW

File: lib/AST/ASTImporter.cpp [hunk 1]
```diff
@@ -3363,6 +3363,30 @@ getFriendCountAndPosition(FriendDecl *FD) {
   return std::make_tuple(FriendCount, *FriendPosition);
 }
 
+DeclContext* getDCOfUnderlyingDecl(FriendDecl *FrD) {
```
Comment (martong) [inline lib/AST/ASTImporter.cpp]: should this be `static`?

Comment (balazske) [reply lib/AST/ASTImporter.cpp]: Yes: I found similar function that are static (`setTypedefNameForAnonDecl`).

File: lib/AST/ASTImporter.cpp [hunk 2]
```diff
@@ -3363,6 +3363,30 @@ getFriendCountAndPosition(FriendDecl *FD) {
...
+        return SpecTy->getAsCXXRecordDecl()->getDeclContext();
+      else if (const auto TypedefTy = dyn_cast<TypedefType>(Ty)) {
+        return TypedefTy->getDecl()->getDeclContext();
+      } else {
+        llvm_unreachable("Unhandled type of friend");
+      }
+    }
+  }
+  // DependentType
+  return nullptr;
```
Comment (balazske) [inline lib/AST/ASTImporter.cpp]: This function can return `Optional` too.

File: lib/AST/ASTImporter.cpp [hunk 3]
```diff
@@ -3363,6 +3363,30 @@ getFriendCountAndPosition(FriendDecl *FD) {
   return std::make_tuple(FriendCount, *FriendPosition);
 }
 
+DeclContext* getDCOfUnderlyingDecl(FriendDecl *FrD) {
+  if (NamedDecl *ND = FrD->getFriendDecl()) {
+    return ND->getDeclContext();
+  }
```
Comment (balazske) [inline lib/AST/ASTImporter.cpp]: Braces can be removed.

File: lib/AST/ASTImporter.cpp [hunk 4]
```diff
@@ -3363,6 +3363,30 @@ getFriendCountAndPosition(FriendDecl *FD) {
...
+    if (!Ty->isDependentType()) {
+      if (const auto *RTy = dyn_cast<RecordType>(Ty))
+        return RTy->getAsCXXRecordDecl()->getDeclContext();
+      else if (const auto *SpecTy = dyn_cast<TemplateSpecializationType>(Ty))
+        return SpecTy->getAsCXXRecordDecl()->getDeclContext();
+      else if (const auto TypedefTy = dyn_cast<TypedefType>(Ty)) {
+        return TypedefTy->getDecl()->getDeclContext();
+      } else {
+        llvm_unreachable("Unhandled type of friend");
+      }
```
Comment (balazske) [inline lib/AST/ASTImporter.cpp]: {} can be removed from here too (at `if` and `else`).

**Comment by martong** [comment]
#624 is related

File: lib/AST/ASTImporter.cpp [hunk 5]
```diff
@@ -3378,6 +3400,21 @@ ExpectedDecl ASTNodeImporter::VisitFriendDecl(FriendDecl *D) {
 
   while (ImportedFriend) {
```
Comment (balazske) [inline lib/AST/ASTImporter.cpp]: Probably it would be better to make this a for-loop, then `continue` is better to use.

Comment (martong) [reply lib/AST/ASTImporter.cpp]: Good idea, thanks.

**Review by balazske — APPROVED** [review]

**Comment by martong** [comment]
https://reviews.llvm.org/D75922

**Pull request merged**